indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedResourceManagementStrategy.java

```diff
@@ -0,0 +1,360 @@
+ * under the License.
+ */
+
+package io.druid.indexing.overlord.autoscaling;
+
+import com.google.api.client.util.Maps;
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.base.Supplier;
+import com.google.common.collect.Collections2;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import com.metamx.emitter.EmittingLogger;
+import io.druid.indexing.common.task.Task;
+import io.druid.indexing.overlord.ImmutableZkWorker;
+import io.druid.indexing.overlord.RemoteTaskRunner;
+import io.druid.indexing.overlord.ZkWorker;
+import io.druid.indexing.overlord.config.RemoteTaskRunnerConfig;
+import io.druid.indexing.overlord.setup.WorkerBehaviorConfig;
+import io.druid.indexing.overlord.setup.WorkerSelectStrategy;
+import io.druid.indexing.worker.Worker;
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ */
+public class PendingTaskBasedResourceManagementStrategy implements ResourceManagementStrategy
+{
+  private static final EmittingLogger log = new EmittingLogger(PendingTaskBasedResourceManagementStrategy.class);
+
+  private final PendingTaskBasedResourceManagementConfig config;
+  private final Supplier<WorkerBehaviorConfig> workerConfigRef;
+  private final ScalingStats scalingStats;
+
+  private final Object lock = new Object();
+  private final Set<String> currentlyProvisioning = Sets.newHashSet();
+  private final Set<String> currentlyTerminating = Sets.newHashSet();
+
+  private DateTime lastProvisionTime = new DateTime();
+  private DateTime lastTerminateTime = new DateTime();
+
+  @Inject
+  public PendingTaskBasedResourceManagementStrategy(
+      PendingTaskBasedResourceManagementConfig config,
+      Supplier<WorkerBehaviorConfig> workerConfigRef
+  )
+  {
+    this.config = config;
+    this.workerConfigRef = workerConfigRef;
+    this.scalingStats = new ScalingStats(config.getNumEventsToTrack());
+  }
+
+  @Override
+  public boolean doProvision(RemoteTaskRunner runner)
+  {
+    Collection<Task> pendingTasks = runner.getPendingTaskPayloads();
+    Collection<ZkWorker> zkWorkers = runner.getWorkers();
+    synchronized (lock) {
+      boolean didProvision = false;
+      final WorkerBehaviorConfig workerConfig = workerConfigRef.get();
+      if (workerConfig == null || workerConfig.getAutoScaler() == null) {
+        log.warn("No workerConfig available, cannot provision new workers.");
+        return false;
+      }
+
+      final List<String> workerNodeIds = workerConfig.getAutoScaler().ipToIdLookup(
+          Lists.newArrayList(
+              Iterables.<ZkWorker, String>transform(
```

Comment:
we seem to be doing this transformation from worker to ids further down as well maybe we can reuse the code?